plugins/cameraserver/src/main/java/edu/wpi/first/shuffleboard/plugin/cameraserver/source/CameraUrlGenerator.java

```diff
@@ -45,31 +51,95 @@ public CameraUrlGenerator(CameraServerSource source) {
     if (frameRate > 0) {
       commands.put("fps", Integer.toString(frameRate));
     }
-    return generateUrls(commands, baseUrls);
+    // Add a special case command for NI Cameras
+    if (source.getName().contains("IMAQdx")) {
+      commands.put("name", source.getName());
+    }
+    return generateUrls(commands, baseUrls, source.getName());
   }
 
   @VisibleForTesting
-  static String[] generateUrls(Map<String, String> commands, String[] baseUrls) { // NOPMD varargs instead of array
+  static String[] generateUrls(Map<String, String> commands, String[] baseUrls,
+                               String cameraName) {
     if (baseUrls == null || baseUrls.length == 0) {
       return new String[0];
     }
     if (commands.isEmpty()) {
       return baseUrls;
     } else {
-      return Arrays.stream(baseUrls)
-          .map(url -> url + toHttpParams(commands))
+      var urls =  Arrays.stream(baseUrls)
+          .map(url -> addHttpParams(url, commands))
+          .filter(url -> url != null)
           .toArray(String[]::new);
+      return urls;
     }
   }
 
   @VisibleForTesting
-  static String toHttpParams(Map<String, String> commands) {
+  static String addHttpParams(String input, Map<String, String> commands) {
     if (commands.isEmpty()) {
-      return "";
+      return input;
+    }
+    // Parse the URI
+    URI uri;
+    try {
+      uri = new URI(input);
+    } catch (URISyntaxException ex) {
+      return input;
     }
-    return commands.entrySet().stream()
-        .map(e -> e.getKey() + "=" + e.getValue())
+
+    String query = uri.getQuery();
+
+    if (query != null) {
+      // Handle the NI special case
```

Comment:
Please add unit tests for NI camera URLs

Reply:
Fixed